File: src/09-lottery/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

function generateDraw(drawType: string): number[] | string {
  let maxNumber: number;
  let maxCount: number;

  switch (drawType) {
    case "five-number draw":
      maxNumber = 90;
      maxCount = 5;
      break;
    case "six-number draw":
      maxNumber = 45;
      maxCount = 6;
      break;
    case "seven-number draw":
      maxNumber = 35;
      maxCount = 7;
      break;
    default:
      return "Error: please choose between 'five-number draw' or 'six-number draw' or 'seven-number draw'.";
  }

  const drawnNumbers: number[] = [];
  while (drawnNumbers.length < maxCount) {
    const randomNumber = Math.floor(Math.random() * maxNumber) + 1;
    if (!drawnNumbers.includes(randomNumber)) {
      drawnNumbers.push(randomNumber);
    }
  }

  drawnNumbers.sort((a, b) => a - b);
  return drawnNumbers;
}

function printDraw(drawType: string): void {
  const draw = generateDraw(drawType);
  if (typeof draw === "string") {
    console.log(draw);
  } else {
    console.log(draw);
  }
}

function main(): void {
  console.log("Lottery draw");
  printDraw("wrong draw-y-x-");
  printDraw("five-number draw");
  printDraw("six-number draw");
  printDraw("seven-number draw");
}

main();

File: src/09-lottery/rust/src/main.rs
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

use rand::prelude::*;

// General function which can draw multiple game types
// These lottery draws are popular in Hungary
// range # count
// 1..90 # 5
// 1..45 # 6
// 1..35 # 7
fn generate_draw(draw_type: &str) -> Result<Vec<i32>, String> {
    match draw_type {
        "five-number draw" | "six-number draw" | "seven-number draw" => {
            let mut drawn_numbers: Vec<i32> = Vec::new();
            let max_number: i32;
            let max_count: i32;

            match draw_type {
                "five-number draw" => {
                    max_number = 90;
                    max_count = 5;
                }
                "six-number draw" => {
                    max_number = 45;
                    max_count = 6;
                }
                "seven-number draw" => {
                    max_number = 35;
                    max_count = 7;
                }
                _ => {
                    max_number = 90;
                    max_count = 5;
                }
            }

            let mut rng: ThreadRng = thread_rng();
            loop {
                let random_number: i32 = rng.gen_range(1..max_number);

                if !drawn_numbers.contains(&random_number) {
                    drawn_numbers.push(random_number);
                }
                if drawn_numbers.len() == max_count as usize {
                    break;
                }
            }
            drawn_numbers.sort();
            Ok(drawn_numbers)
        }
        _ => {
            Err(String::from("Error: please choose between 'five-number draw' or 'six-number draw' or 'seven-number draw'."))
        }
    }
}

fn print_draw(draw_type: &str) {
    let draw = generate_draw(draw_type);
    match draw {
        Ok(val) => {
            let d = format!("{val:?}");
            println!("{d}");
        }
        Err(e) => println!("{e}"),
    }
}

fn main() {
    println!("Lottery draw");
    print_draw("wrong draw-y-x-");
    print_draw("five-number draw");
    print_draw("six-number draw");
    print_draw("seven-number draw");
}
